### PR TITLE
[TypeChecker] `setInvalid()`s cleanup

### PR DESCRIPTION
**Pull request by HassanElDesouky**
In this PR I removed all of the `setInvalid()`s from the TypeCheckDeclPrimary file and modified all of the test messages as discussed in both of these comments:
1. https://github.com/apple/swift/pull/30824#issuecomment-613774410
2. https://github.com/apple/swift/pull/30824#issuecomment-613781036

All of the tests pass locally *on Ubuntu*, however, there's only one test that I wasn't able to make it work and it's [this test in Parse/enum.swift](https://github.com/apple/swift/blob/master/test/Parse/enum.swift#L367)

I found out that the compiler crashes because of removing only the `setInvalid()` [in the TypeCheckDeclPrimary at L721](https://github.com/apple/swift/blob/master/lib/Sema/TypeCheckDeclPrimary.cpp#L721) If I didn't remove it this test will pass, however, the diagnostic will change for all of the other tests of course. I was thinking of removing this test which is causing the headache but of course, I'm sure that this is not the correct way to do it :)

How can I solve this?

cc @CodaFi @xedin @LucianoPAlmeida 

#30824 aka #31037 Follow-up.

### CODE REVIEW

**Review by CodaFi — CHANGES_REQUESTED** [review]
Please only limit the removal to the part of TypeCheckDeclPrimary we discussed earlier. Some of these tests have regressed in diagnostic quality because of this change.

**Comment by HassanElDesouky** [comment]
> Please only limit the removal to the part of TypeCheckDeclPrimary we discussed earlier. Some of these tests have regressed in diagnostic quality because of this change.

I'm sorry, I thought you mean remove all of the setInvalid()s occurrences in the file. Anyways, I'll get back to it. However, what should I do about https://github.com/apple/swift/blob/master/test/Parse/enum.swift#L367?

**Comment by CodaFi** [comment]
If there's a crash because of a structural problem in the type checker, we need to address that at a more fundamental level. That may involve fixing up TypeCheckPattern to be more tolerant of invalid code.

File: test/Parse/operator_decl_designated_types.swift [hunk 1]
```diff
@@ -33,11 +33,15 @@ infix operator **+> : MediumPrecedence, UndeclaredProtocol
 // expected-error@-1 {{use of undeclared type 'UndeclaredProtocol'}}
 
 prefix operator *+*> : MediumPrecedence
+// expected-error@-1 {{use of undeclared type 'MediumPrecedence'}}
```
Comment (LucianoPAlmeida) [inline test/Parse/operator_decl_designated_types.swift]: I think this is a regression, `MediumPrecedence` is declared right above on line 7. Right?

Comment (HassanElDesouky) [reply test/Parse/operator_decl_designated_types.swift]: Yes, I think so.

**Comment by HassanElDesouky** [comment]
Because of all of the regression that happened, I decided to remove only the `setInvalid()`s in this function, for now. https://github.com/apple/swift/blob/426583523e063be1ed75027044f56873b7b73183/lib/Sema/TypeCheckDeclPrimary.cpp#L483

The next thing was this test case which was causing a compiler crash
> there's only one test that I wasn't able to make it work and it's this test in Parse/enum.swift https://github.com/apple/swift/blob/426583523e063be1ed75027044f56873b7b73183/test/Parse/enum.swift#L367

As @CodaFi pointed me, the problem was in the `TypeCheckPattern` file. 

**Why the compiler is crashing?**
The compiler crash was caused by this assertion at 
https://github.com/apple/swift/blob/426583523e063be1ed75027044f56873b7b73183/lib/Sema/TypeCheckPattern.cpp#L98

*I think* that's because at  
https://github.com/apple/swift/blob/426583523e063be1ed75027044f56873b7b73183/lib/Sema/TypeCheckDeclPrimary.cpp#L721
was setting the redeclaration as invalid so when it goes into this line it will [go into the `if` statement ](https://github.com/apple/swift/blob/master/lib/Sema/TypeCheckPattern.cpp#L85) and the loop will skip over it. However now, after deleting the `setInvalid` line in TypeCheckDeclPrimary this [`if` statement ](https://github.com/apple/swift/blob/master/lib/Sema/TypeCheckPattern.cpp#L85) won't be `true` so it won't go into it and thus, fail in the assertion at [L98 in TypeCheckPattern](https://github.com/apple/swift/blob/426583523e063be1ed75027044f56873b7b73183/lib/Sema/TypeCheckPattern.cpp#L98)

**My solution** 
My solution was just to delete the assertion line and everything went fine. 

**Diagnostics regressions**
I think there's a regression in this file, however, I don't know how to fix it. 
https://github.com/apple/swift/blob/2afd1b0027798e7c661c59771dbf4652fe119d00/test/TypeCoercion/overload_noncall.swift#L55

Sorry for the long summery, but I just wanted to share what I came to. :) 

cc @CodaFi , @LucianoPAlmeida , @xedin

File: lib/Sema/TypeCheckPattern.cpp [hunk 2]
```diff
@@ -94,8 +94,6 @@ filterForEnumElement(DeclContext *DC, SourceLoc UseLoc,
     }
 
     if (auto *oe = dyn_cast<EnumElementDecl>(e)) {
-      // Ambiguities should be ruled out by parsing.
-      assert(!foundElement && "ambiguity in enum case name lookup?!");
```
Comment (CodaFi) [inline lib/Sema/TypeCheckPattern.cpp:98]: We shouldn't just remove this assertion. This amounts to a non-trivial behavior change in this function - it's now going to break ambiguities by selecting the last returned result from a lookup. The trouble here isn't the assert, it's the 

```
    if (e->isInvalid()) {
      continue;
    }
```

check guarding the filtering behavior. That condition was correct in 2016 when this was written, but is not great now that we've stabilized the computation of interface types and defined `isInvalid` in terms of it.

Comment (CodaFi) [reply lib/Sema/TypeCheckPattern.cpp:98]: I *think* this means this code just needs to become tolerant of duplicates. Quite what that means exactly I'm not sure - I need to go page all of this back in.

**Review by CodaFi — CHANGES_REQUESTED** [review]
We're getting closer. I think you may have fallen down a bit of a rabbit hole - but that's not unexpected given how old this code is!

**Comment by CodaFi** [comment]
Regardless, let's see what happens, eh?

@swift-ci test

**Comment by CodaFi** [comment]
@swift-ci test source compatibility

**Comment by CodaFi** [comment]
@swift-ci test

**Comment by CodaFi** [comment]
@swift-ci test source compatibility

**Comment by CodaFi** [comment]
Huh, 

@swift-ci test source compatibility Debug

File: lib/Sema/TypeCheckPattern.cpp (hunk 2 above)
Comment (HassanElDesouky) [reply lib/Sema/TypeCheckPattern.cpp:98]: Does that mean I might need to check for redeclarations in TypeCheckPattern?

Comment (HassanElDesouky) [reply lib/Sema/TypeCheckPattern.cpp:98]: Thank you so much for helping me. And [Ramadan Kareem](https://www.thesun.co.uk/news/6211919/ramadan-kareem-mubarak-fasting-meaning/)! 🙏 :)

**Comment by shahmishal** [comment]
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

**Pull request closed without merging**